### ISSUE TITLE
Memory spikes in model creation

### ISSUE THREAD

**Issue by koen-vg** [PyPSA/PyPSA#1051]
### Version Checks (indicate both or one)

- [X] I have confirmed this bug exists on the lastest [release](https://github.com/pypsa/pypsa/releases) of PyPSA.

- [ ] I have confirmed this bug exists on the current [`master`](https://github.com/pypsa/pypsa/tree/master) branch of PyPSA.

### Issue Description

I have previously brought up the problem of memory spikes during model creation, arising specifically in the context of myopic foresight optimisations with pypsa-eur, see https://github.com/PyPSA/linopy/issues/248 and https://github.com/PyPSA/pypsa-eur/pull/1056. I thought it would be useful to open an issue in this repo specifically about memory usage during model building in pypsa, as I've at least identified one specific place in the code where memory spikes, and it looks like maybe someone with more time/expertise than I have could make some improvements.

To recap, this is what memory usage typically looks like during model building with pypsa-eur under myopic foresight:
![bilde](https://github.com/user-attachments/assets/e81cfebb-2846-4b7c-a6c6-231a433e80dc)
The graph flattening out at the end is the start of gurobi taking over and continues for much longer. So for the vast majority of time, this model needed under 22GB, but during model building this spiked to almost 40GB in this example. This is very inconvenient when using pypsa-eur (or any pypsa-based workflow) in a SLURM set-up where jobs must be assigned hard memory-caps.

As pointed out in https://github.com/PyPSA/linopy/issues/248#issuecomment-2124376159, the first peak occurs when defining nodal balance constraints, and this is what I think it might be possible to address somehow at the pypsa-level. Playing around with some memory monitoring and breakpoints, I have narrowed down the _peak_ of the first memory spike to a single line: https://github.com/PyPSA/PyPSA/blob/6bc7f2dbf1e831094d8b81998c3cabac2b44fa06/pypsa/optimization/constraints.py#L594
This merge operation in particular causes the several (hard to see here, but four) individual peaks, each going to 37-39GB. I believe that the expression building immediately preceding the linked line causes the run-up to this succession of peaks.

Shouldn't it somehow be possible to do the expression-building in a more careful way? Obviously the amount of memory consumption is "too much" in the sense that the full model does not in fact take so much memory to represent.

Unfortunately I personally don't have the time to look into this any more right now. But since I spent a bit of time this morning, I thought I should at least note down the results of my investigation, in case anyone else is willing to follow this up.

### Reproducible Example

```python
# Run pypsa-eur with myopic foresight and relatively high spacial and temporal resolution, e.g. 60 nodes, 3-hourly resolution.
```

### Expected Behavior

PyPSA should, while building a model, never use more memory than it takes to actually solve the model.

### Installed Versions

<details>

Version: 0.30.2

</details>

**Comment by MaykThewessen** [comment]
Is there already a solution for this?

would like to run PyPSA Europe-sector coupled on a 24GB or 64GB Macbook M4 Pro

**Comment by koen-vg** [comment]
I don't think there's a solution to this particular problem yet, but in general you might just have to downsize your problem? I believe that generally when you have a model with sufficiently high resolution that this becomes a problem (i.e. memory spike above 24GB), the solving also tends to take a good number of hours; not very convenient on a laptop anyway. But you can get small enough problems from pypsa-eur by selecting an appropriate spatio-temporal resolution and get acceptable results.

**Comment by lkstrp** [comment]
We also couldn't replicate this properly. All the profiles done for #1154 show a lower peak than the solver needs.

But maybe we should try harder. ~@koen-vg Could you share the config?~ We'll take a new one

**Comment by FabianHofmann** [comment]
I am quite confident with the slicing logic in linopy introduced in v0.4.3/v0.4.4 there should not be memory spikes anymore. but would be curious if @koen-vg can confirm this

**Comment by MaykThewessen** [comment]
I Forked the PyPSA Form Energy Europe model from OET
https://github.com/open-energy-transition/form-energy-storage 


That required me to downgrade Linopy to an older version, because of compatibility issues

And PyPSA 0.32 with Linopy 0.3?

#dependencies:
- python>=3.8
- pip

#Inhouse packages
- pypsa>=0.32
- atlite>=0.2.9
- linopy
- powerplantmatching>=0.5.15,<0.6

How can we use the latest PyPSA and Linopy packages

Without compatibility issues with a repository from 1.5y ago?

How to change a forked script to be compatible with a newer one

**Comment by lkstrp** [comment]
Looks like @FabianHofmann is right and we can close this:

<img width="3571" height="2365" alt="Image" src="https://github.com/user-attachments/assets/aadd842b-0132-44b6-a55b-7409b3fb1090" />

@koen-vg Reopen the issue if you can't confirm

@MaykThewessen 
You can either fix the compatibility issues yourself or merge the latest pypsa-eur branch into your fork. If you have already switched from pyomo to linopy, the former option should be manageable. But this is not linked to the issue here. Create an issue in the linked repository to ask if OET plans to synchronise their fork.

**Comment by koen-vg** [comment]
I agree, I just tested a little bit from my side and as far as I can see, the model building memory spikes are gone. Great work, thanks!
